### ISSUE TITLE
Upload to Hackage

### ISSUE THREAD

**Issue by eborden** [frontrowed/faktory_worker_haskell#30]
What's left? Do we need anything else to go to `hackage` and make an official release?

**Comment by eborden** [comment]
I ran a `neil check` and got these
https://github.com/ndmitchell/neil

 - [ ] Warning: 'ghc-options: -O0' is not needed. Use the --disable-optimization
 - [ ] configure flag.
 - [ ] Warning: The following errors will cause portability problems on other
 - [ ] environments:
 - [ ] Warning: No 'synopsis' or 'description' field.
 - [ ] Warning: The dependency 'build-depends: base' does not specify an upper bound
 - [ ] on the version number. Each major release of the 'base' package changes the
 - [ ] API in various ways and most packages will need some changes to compile with
 - [ ] it. The recommended practise is to specify an upper bound on the version of
 - [ ] the 'base' package. This ensures your package will continue to build when a
 - [ ] new major version of the 'base' package is released. If you are not sure what
 - [ ] upper bound to use then use the next major version. For example if you have
 - [ ] tested your package with 'base' version 4.5 and 4.6 then use 'build-depends:
 - [ ] base >= 4.5 && < 4.7'.
 - [ ] Warning: Hackage would reject this package.

**Comment by pbrisbin** [comment]
> What's left?

Personally, I just wanted to see a fair amount of production usage and iron out any rough edges. There's no technical way to know when that's "done", but I feel pretty good about things at the moment and would not be against cutting a 1.0 here.

That checklist seems reasonable to me.

---

Has there been any structured discussion about how we version our OSS libraries? For my own libraries, I use Semantic Versioning (because that's what I know and prefer) but with a prefix that makes it also PVP-compliant (to be a good Hackage citizen): `0.{MAJOR}.{MINOR}.{PATCH}`. I'm happy to use that, but am not pushing for it. So long as whatever we do use is logical and documented, I'm good.

**Comment by eborden** [comment]
Yeah, I'd prefer to use the `PVP` for hackage libraries. I don't see any major gains in going against the grain there. `SemVer` is fine and dandy, but hackage prefers the `PVP`.

Side note, I've also utilized the first value before. I tend to think of it as `{EPOCH}.{BREAKING}.{MINOR}.{PATCH}`.

For example, if we were to view `python` through the `PVP`, the jump from version 2 to 3 was most definitely an `EPOCH`.

**Comment by pbrisbin** [comment]
> Yeah, I'd prefer to use the PVP for hackage libraries. I don't see any major gains in going against the grain there. SemVer is fine and dandy, but hackage prefers the PVP.

Sorry if you already understood this, but that comment makes me think you might've missed it: my proposed scheme is SemVer **and** PVP -- that is its main strength, and why I use it.

> I've also utilized the first value before. I tend to think of it as {EPOCH}.{BREAKING}.{MINOR}.{PATCH}.

Exactly how I think about it too; I'm just saying start `EPOCH` at `0` and what you call `BREAKING`, I call `MAJOR` to align with SemVer terminology.

---

I can't really tell, are you against my proposed scheme? Suggesting something else?

**Comment by eborden** [comment]
I did not understand that you were attempting to combine the two. If we use `SemVer` we should use `SemVer`. If we use the `PVP` we should use the `PVP`. Trying to combine them into some kind of Frankenstein seems like it will just cause pain.

**Comment by eborden** [comment]
Also `PVP` calls it `MAJOR`, I just think `BREAKING` is easier to understand (`MAJOR.MAJOR.MINOR.PATCH`).

**Comment by eborden** [comment]
The PVP decision tree.

![image](https://user-images.githubusercontent.com/545655/47508139-23ca5c00-d839-11e8-8db7-bb4206d947bc.png)

**Comment by pbrisbin** [comment]
> seems like it will just cause pain

Anecdotal, but it has not for me, at all. It's very simple. It is really just SemVer, but having the `EPOCH` gets the `A.B.C` of that decision tree to line up so you get the same result if you thought about it as PVP (major bumps `B`, minor bumps `C`).

Happy to just follow that decision tree each release, if that's what you're proposing.

**Comment by eborden** [comment]
If that conceptual framework works for you :+1:

Just to be super explicit. We agree that the version scheme looks like:
```
MAJOR.MAJOR.MINOR.PATCH
```
where
1. Breaking changes may only happen in `MAJOR`
2. New (non-breaking) API additions happen in `MINOR`
3. Bug fixes that don't effect the API happen in `PATCH`

**Comment by pbrisbin** [comment]
I love super explicit. What you said is accurate, :+1:

Just want to pint down two minor details:

- You're leaving it vague when/if to bump `MAJOR` (1) vs `MAJOR` (2), that's left to figure out case-by-case? Any additional guidance worth writing down?
- We're using the decision tree to determine "breaking", right?

**Comment by eborden** [comment]
> You're leaving it vague when/if to bump MAJOR (1) vs MAJOR (2), that's left to figure out case-by-case? 
> Any additional guidance worth writing down?
We could define a percentage of API breakage? If there is a major reworking of sorts. It is very easy to produce `API` breakage through just adding an argument to a function, so `MAJOR (2)` is not hard to increment. `MAJOR (1)` is for signalling a significant amount of breakage. A rewrite, a fundamental shift in the API, etc. For example in Faktory, I'd consider a shift from RocksDB to Redis to be a `MAJOR (1)`, where is a small breaking protocol change would be `MAJOR (2)`. I'm not sure how best to formalize that.

> We're using the decision tree to determine "breaking", right?
:+1:

**Comment by pbrisbin** [comment]
> I'm not sure how best to formalize that.

IMO, by doing this:

```
EPOCH.MAJOR.MINOR.PATCH
```

where

1. A rewrite or major shift in API happens in `EPOCH`
1. Breaking changes may only happen in `MAJOR` (or `EPOCH`)
1. New (non-breaking) API additions happen in `MINOR`
1. Bug fixes that don't effect the API happen in `PATCH`

Which is just what I suggested originally...

**Comment by eborden** [comment]
:+1: to this more formal definition. It seems we were mostly in violent agreement.

**Comment by pbrisbin** [comment]
@eborden I think we're ready to go ahead with our 1.0 here. There are just two questions (IMO):

1. Is the team good on the versioning policy we came up with, and are we starting as `1.0.0.0`, `1.1.0.0`, or `0.1.0.0`? All of these could qualify as "1.0" to me.
1. Should we upload via a shared Hackage account or as one of ourselves?

I would prefer a shared account, all things being equal, but I'm concerned that `stack upload` is too nice a UX for uploading with a consistent set of credentials. Having to use specific credentials between personal and Freckle packages might be a poor UX -- I don't know this, I just suspect it.

**Comment by eborden** [comment]
> Is the team good on the versioning policy we came up with, and are we starting as 1.0.0.0, 1.1.0.0, or 0.1.0.0? All of these could qualify as "1.0" to me.

I'm cool with `1.0.0.0` for marketing reasons.

> Should we upload via a shared Hackage account or as one of ourselves?

I think it would be good to add a shared freckle account to the maintainers list, but also include our own hackage accounts for ease of use.

**Comment by pbrisbin** [comment]
SGTM -- I didn't realize we could have multiple maintainers like that.
